pkg/cli/cli.go

```diff
@@ -0,0 +1,498 @@
+	// (ex. "go.kubebuilder.io") keys or both, their values may need to be
+	// resolved to known plugins by key.
+	plugins := c.pluginsFromOptions[c.projectVersion]
+	switch {
+	case c.cliPluginKey != "":
+		// Filter plugin by keys passed in CLI.
+		c.resolvedPlugins, err = resolvePluginsByKey(plugins, c.cliPluginKey)
+	case c.configured && projectConfig.IsV3():
+		// All non-v1 configs must have a layout key. This check will help with
+		// migration.
+		if projectConfig.Layout == "" {
+			return fmt.Errorf("config must have a layout value")
+		}
+		// Filter plugin by config's layout value.
+		c.resolvedPlugins, err = resolvePluginsByKey(plugins, projectConfig.Layout)
+	default:
+		// Use the default plugins for this project version.
+		c.resolvedPlugins = c.defaultPluginsFromOptions[c.projectVersion]
+	}
+	if err != nil {
+		return err
+	}
+
+	c.cmd = c.buildRootCmd()
+
+	// Add extra commands injected by options.
+	for _, cmd := range c.extraCommands {
+		for _, subCmd := range c.cmd.Commands() {
+			if cmd.Name() == subCmd.Name() {
+				return fmt.Errorf("command %q already exists", cmd.Name())
+			}
+		}
+		c.cmd.AddCommand(cmd)
+	}
+
+	// Write deprecation notices after all commands have been constructed.
+	for _, p := range c.resolvedPlugins {
+		if d, isDeprecated := p.(plugin.Deprecated); isDeprecated {
+			fmt.Printf(noticeColor, fmt.Sprintf("[Deprecation Notice] %s\n\n",
+				d.DeprecationWarning()))
+		}
+	}
+
+	return nil
+}
+
+// parseBaseFlags parses the command line arguments, looking for flags that
+// affect initialization of a cli. An error is returned only if an error
+// unrelated to flag parsing occurs.
+func (c *cli) parseBaseFlags() error {
+	// Create a dummy "base" flagset to populate from CLI args.
+	fs := pflag.NewFlagSet("base", pflag.ExitOnError)
+	fs.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
+
+	var help bool
+	// Set base flags that require pre-parsing to initialize c.
+	fs.BoolVarP(&help, helpFlag, "h", false, "print help")
+	fs.StringVar(&c.projectVersion, projectVersionFlag, c.defaultProjectVersion, "project version")
+	fs.StringVar(&c.cliPluginKey, pluginsFlag, "", "plugins to run")
+
+	// Parse current CLI args outside of cobra.
+	err := fs.Parse(os.Args[1:])
+	// User needs *generic* help if args are incorrect or --help is set and
+	// --project-version is not set. Plugin-specific help is given if a
+	// plugin.Context is updated, which does not require this field.
+	c.doGenericHelp = err != nil || help && !fs.Lookup(projectVersionFlag).Changed
+	c.cliPluginKey = strings.TrimSpace(c.cliPluginKey)
+
+	return nil
+}
+
+// validate validates fields in a cli.
+func (c cli) validate() error {
+	// Validate project versions.
+	if err := validation.ValidateProjectVersion(c.defaultProjectVersion); err != nil {
+		return fmt.Errorf("failed to validate default project version %q: %v", c.defaultProjectVersion, err)
+	}
```

Comment:
Nit/question: At this point, do we use `c.defaultProjectVersion` anymore? It looks like we've already populated `c.projectVersion` from `c.defaultProjectVersion` in `initialize()`.

When `projectVersion` and `defaultProjectVersion` are the same, we're already validating `projectVersion`, so this is duplicative in that case. When they're different, the user doesn't care that `defaultProjectVersion` is invalid because it isn't being used, right?

If I'm understanding correctly, the default project version is defined by the CLI owner, not the user. So this error would not be actionable or solvable by a user, right? Should we `panic` instead with an error message that makes it more obvious that there's an issue for maintainers to fix?

Reply:
Agreed, I'll improve the error message.

Reply:
Also it makes more sense to move these checks into their `Option` functions, since they're not user input.